src/defrag.c

```diff
@@ -604,15 +612,29 @@ DefragInsertFrag(ThreadVars *tv, DecodeThreadVars *dtv, DefragTracker *tracker,
     tracker->timeout.tv_sec = p->ts.tv_sec + tracker->host_timeout;
     tracker->timeout.tv_usec = p->ts.tv_usec;
 
-    Frag *prev = NULL, *next;
+    Frag *prev = NULL, *next = NULL;
     int overlap = 0;
     ltrim = 0;
-    if (!TAILQ_EMPTY(&tracker->frags)) {
-        TAILQ_FOREACH(prev, &tracker->frags, next) {
+
+    if (!RB_EMPTY(&tracker->fragment_tree)) {
+        Frag key = {
+            .offset = frag_offset - 1,
+        };
+        next = RB_NFIND(IP_FRAGMENTS, &tracker->fragment_tree, &key);
+        if (next == NULL) {
+            prev = RB_MIN(IP_FRAGMENTS, &tracker->fragment_tree);
+            next = IP_FRAGMENTS_RB_NEXT(prev);
+        } else {
+            prev = IP_FRAGMENTS_RB_PREV(next);
+            if (prev == NULL) {
+                prev = next;
+                next = IP_FRAGMENTS_RB_NEXT(prev);
+            }
+        }
+        while (prev != NULL) {
```

Comment:
I think this could be much simplified. In the TCP case I've made sure that the compare func primarily looks at the seq (offset here) and secondary at the length. This means that I can simply RB_INSERT. Only if exactly equal it rejects the insert. In other cases, you can take the inserted frag, check RB_PREV to see if that overlaps us and check RB_NEXT to see if we overlap that. This way no looping is required.

Reply:
Whats the case for having exact duplicates in the tree?

Reply:
In the "LAST" policy, the data received last takes precedence. So it may mean removing a fragment that matches it in offset and length. We don't actually remove it from the tree, it just gets flagged as skipped and the new fragment inserted.

I think the insert could be simplified by moving some of this logic into the compare function, but that would require relaying the personality through, and moving some of the logic. I didn't want to go that far yet.

Reply:
But in this case it would be enough to remove the "old" duplicate and replace it by the new frag for that exact offset+len right?

Reply:
It could. Not saying it couldn't be done, but this would change the lookup based on what personality is being used? Something I was trying hard to avoid for this run - not touching the personality code.

With "last", if only a portion overlaps, we remove that data from the prior packet. So this lookup catches both of those cases with having the personality code in one section.

Reply:
I'm fine with keeping it this way for now.

Wondering though, can we currently insert large amounts of duplicate frags?

Reply:
(currently as in this PR)

Reply:
Yes, with the last policy this would be possible. I have a fix for this now as well. When inserting a new fragment, if the previous has been marked as skip, or has been trimmed to the point of being remove, remove it from the tree then instead of hanging onto it.